test/with_dummyserver/test_https.py

```diff
@@ -945,6 +947,89 @@ def test_alpn_default(self) -> None:
             r = pool.request("GET", "/alpn_protocol", retries=0)
             assert r.status == 200
             assert r.data.decode("utf-8") == util.ALPN_PROTOCOLS[0]
+            assert (
+                r.data.decode("utf-8") == {"h11": "http/1.1", "h2": "h2"}[http_version]
+            )
+
+    def test_http2_probe_result_is_cached(self, http_version: str) -> None:
+        from urllib3.connection import _HTTP2_PROBE_CACHE as http2_probe_cache
+
+        assert http2_probe_cache.values() == {}
+
+        with HTTPSConnectionPool(
+            self.host,
+            self.port,
+            ca_certs=DEFAULT_CA,
+        ) as pool:
+            r = pool.request("GET", "/alpn_protocol", retries=0)
+            assert r.status == 200
+
+        if http_version == "h2":
+            # This means the probe was successful.
+            assert http2_probe_cache.values() == {(self.host, self.port): True}
+        else:
+            # This means the probe wasn't attempted, otherwise would have a value.
+            assert http_version == "h11"
+            assert http2_probe_cache.values() == {}
+
+    @pytest.mark.xfail(reason="Hypercorn always supports both HTTP/2 and HTTP/1.1")
```

Comment:
Do we want a fixture where it only supports one?

Reply:
The mixed case is fine for the future, maybe this can be opened as a separate issue to allow for the server to be launched in one of two modes?